### PR TITLE
Feature/temporal evaluator

### PR DESCRIPTION
**Pull request by tajinder-txstate**
#663 
-added temporal evaluator
-added temporal evaluator test
-added simulate command to run rvaluator
-updated binary rating dao with binary search feature

### CODE REVIEW

File: lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+              .setDefault("ratings.pack")
+              .help("Packed Rating File");
+        parser.addArgument("-o", "--output-file")
+              .type(File.class)
+              .metavar("FILE")
+              .setDefault("prediction.csv")
+              .help("write predicted score to FILE");
+        parser.addArgument("-r", "--rebuild-period")
+              .type(Long.class)
+              .metavar("REBUILD PERIOD IN SECONDS")
```
Comment (mdekstrand) [inline lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java]: Metavar should be short, like `SECONDS`.

File: lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+        parser.addArgument("config")
+              .type(File.class)
+              .nargs("+")
+              .metavar("CONFIG")
+              .help("load algorithm configuration from CONFIG");
+    }
+
+    @Override
+    public String getName() {
+        return "simulate-temporal-evaluator";
```
Comment (mdekstrand) [inline lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java]: Let's make the name `simulate`.

File: lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+
+/**
+ * @author <a href="http://www.grouplens.org">GroupLens Research</a>
+ */
+@AutoService(Command.class)
+public class Simulate implements Command {
+    private final Logger logger = LoggerFactory.getLogger(Simulate.class);
+
+    public void configureArguments(ArgumentParser parser) {
+        parser.description("Simulates Temporal Evaluator and write item's predicted score to file.");
```
Comment (mdekstrand) [inline lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java]: Let's describe it as 'Simulates a recommender over time'

File: lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+              .metavar("FILE")
+              .setDefault("prediction.csv")
+              .help("write predicted score to FILE");
+        parser.addArgument("-r", "--rebuild-period")
+              .type(Long.class)
+              .metavar("REBUILD PERIOD IN SECONDS")
+              .help("Rebuild Period for next build");
+        parser.addArgument("config")
+              .type(File.class)
+              .nargs("+")
```
Comment (mdekstrand) [inline lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java]: If you remove this `nargs` line, that should make it take one config file.

File: lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+
+        public File getInputFile() {
+            return options.get("input_file");
+        }
+
+        public File getOutputFile() {
+            return options.get("output_file");
+        }
+
+        public List<File> getConfigFiles() {
```
Comment (mdekstrand) [inline lenskit-cli/src/main/java/org/lenskit/cli/commands/Simulate.java]: If you change this to `File getConfigFile()`, then we'll just have 1 config file.

File: lenskit-cli/src/test/groovy/org/lenskit/cli/baseline.groovy [hunk 6]
```diff
@@ -0,0 +1,30 @@
+/*
```
Comment (mdekstrand) [inline lenskit-cli/src/test/groovy/org/lenskit/cli/baseline.groovy]: This file should not be committed. It is just for your testing.

File: lenskit-cli/src/test/groovy/org/lenskit/cli/item-item-eval.groovy [hunk 7]
```diff
@@ -0,0 +1,51 @@
+/*
```
Comment (mdekstrand) [inline lenskit-cli/src/test/groovy/org/lenskit/cli/item-item-eval.groovy]: Same with this file.

File: lenskit-core/src/main/java/org/lenskit/data/packed/BinaryRatingDAO.java [hunk 8]
```diff
@@ -453,7 +453,7 @@ private void readObject(ObjectInputStream in) throws IOException, ClassNotFoundE
             itemTable = (BinaryIndexTable) in.readObject();
             limitIndex = in.readInt();
             limitTimestamp = in.readLong();
-
+;
```
Comment (mdekstrand) [inline lenskit-core/src/main/java/org/lenskit/data/packed/BinaryRatingDAO.java]: This looks like a stray change, should be deleted.

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+        return this;
+    }
+
+    /**
+     * @param time The time to rebuild
+     * @param unit Unit of time set
+     * @return Itself for  method chaining
+     */
+    public TemporalEvaluator setRebuildPeriod(Long time, TimeUnit unit) {
+        rebuildPeriod = unit.toSeconds(time);
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: Perfect!

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 10]
```diff
@@ -0,0 +1,225 @@
...
+        //file headers
+        tlb.addColumn("User")
+           .addColumn("Item")
+           .addColumn("Rating")
+           .addColumn("Timestamp")
+           .addColumn("Prediction")
+           .addColumn("TARMSE");
+
+        TableLayout tl = tlb.build();
+        tableWriter = CSVWriter.open(predictOutputFile, tl, CompressionMode.AUTO);
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: This should be immediately before the outer `try`, ideally converted to a `try`-with-resources block.

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 11]
```diff
@@ -0,0 +1,225 @@
...
+                LenskitConfiguration config = new LenskitConfiguration();
+                config.addComponent(limitedDao);
+                if (r.getTimestamp() - buildTime >= rebuildPeriod) {
+                    buildTime = r.getTimestamp();
+                    lre = LenskitRecommenderEngine.newBuilder()
+                                                  .addConfiguration(algorithm.getConfigurations().get(0))
+                                                  .addConfiguration(config, ModelDisposition.EXCLUDED)
+                                                  .build();
+                }
+                recommender = lre.createRecommender(config);
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: We can refactor this code so that it is all in the `if`, and so that we only create a recommender once for a block of ratings with the same timestamp. Please do that.

Also, when testing whether we need to build a new engine, we should explicitly test if `lre` is null to always catch the initial condition.

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 12]
```diff
@@ -0,0 +1,225 @@
...
+        TableLayout tl = tlb.build();
+        tableWriter = CSVWriter.open(predictOutputFile, tl, CompressionMode.AUTO);
+
+        List<Rating> ratings = ObjectStreams.makeList(dataSource.streamEvents(Rating.class, SortOrder.TIMESTAMP));
+        BinaryRatingDAO limitedDao = dataSource.createWindowedView(0);
+        LenskitRecommenderEngine lre;
+        Recommender recommender;
+        double sse = 0;
+        int n = 0;
+        lre = null;
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: Please move this initialization to be with the engine declaration.

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 13]
```diff
@@ -0,0 +1,225 @@
...
+                Result predictionResult = recommender.getRatingPredictor().predict(r.getUserId(), r.getItemId());
+                double predict = Double.NaN;
+                if (predictionResult != null) {
+                    predict = predictionResult.getScore();
+                }
+                //calculates time averaged RMSE
+                if (!Double.isNaN(predict)) {
+                    double err = predict - r.getValue();
+                    sse += err * err;
+                    n++;
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: Perfect.

File: lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java [hunk 14]
```diff
@@ -0,0 +1,225 @@
...
+                if (n > 0) {
+                    rmse = sqrt(sse / n);
+                }
+                //writes the Prediction Score and TARMSE on file
+                tableWriter.writeRow(r.getUserId(), r.getItemId(), r.getValue(), r.getTimestamp(), predict, rmse);
+            }
+        } finally {
+            tableWriter.close();
+        }
+        return this;
```
Comment (mdekstrand) [inline lenskit-eval/src/main/java/org/lenskit/eval/temporal/TemporalEvaluator.java]: No need to return `this` from the execute method.

**Comment by mdekstrand** [comment]
High-level: looks good!

There are some small fixes in documentation and code layout, but I don't see any logic bugs.

We also need to remove the extra files, and we need to remove them retroactively with a rebase (we don't want `ratings.pack` in the repository).

**Comment by mdekstrand** [comment]
I am going to manually merge this code with a thing called a *rebase* that will retroactively edit the history and permanently remove the erroneous files

Once that is done, you will need to delete your branch and just work from 'master'. Merging or pulling will **not** work.

**Comment by mdekstrand** [comment]
Superseded by #785.

**Comment by mdekstrand** [comment]
@tajinder-txstate This is now merged in #785, good work!
